fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,524 @@
+}
+
+impl<'a> Hash for ArgumentInfo<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+impl<'a> Borrow<str> for ArgumentInfo<'a> {
+    fn borrow(&self) -> &str {
+        self.name
+    }
+}
+
+#[derive(Default)]
+pub struct ArgParser<'a> {
+    args_matches: HashSet<ArgumentInfo<'a>>,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(&mut self, arg_info: ArgumentInfo<'a>) {
+        self.args_matches.insert(arg_info);
+    }
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    // Used for a general validation of the arguments provided by user.
+    fn check_conflicts(
+        &self,
+        argument_info: &ArgumentInfo<'a>,
+        trimmed_args: &[&str],
+        index: usize,
+        args: &[String],
+    ) -> Result<()> {
+        // Check if the argument at `index` position was not provided more than once.
+        if self
+            .args_matches
+            .iter()
+            .any(|x| x.name == argument_info.name && x.user_value.is_some())
+        {
+            return Err(Error::DuplicatedArgument(argument_info.name.to_string()));
+        }
+        // If the argument at `index` position takes a value, check if there was provided
+        // one and only one value for that argument.
+        if argument_info.takes_value {
+            if let Some(arg_value) = args.get(index + 1) {
+                if arg_value.starts_with("--")
+                    && trimmed_args.contains(&arg_value.trim_start_matches("--"))
+                {
+                    return Err(Error::MissingValue(argument_info.name.to_string()));
+                }
+                if let Some(arg) = args.get(index + 2) {
+                    if !arg.starts_with("--") {
+                        return Err(Error::UnexpectedArgument(arg.clone()));
+                    }
+                }
+            } else {
+                return Err(Error::MissingValue(argument_info.name.to_string()));
+            }
+        } else {
+            // If the argument at `index` position is a flag, check if there wasn't provided
+            // a value for that argument.
+            if let Some(arg) = args.get(index + 1) {
+                if !arg.starts_with("--") {
+                    return Err(Error::UnexpectedArgument(arg.clone()));
+                }
+            }
+        }
+        // For the arguments that require a specific argument to be also present in the list
+        // of arguments from user, search for that argument.
+        if let Some(arg_name) = argument_info.requires {
+            if !trimmed_args.contains(&arg_name) {
+                return Err(Error::MissingArgument(arg_name.to_string()));
+            }
+        }
+        Ok(())
+    }
+
+    // Populates `args_matches` field of ArgParser.
+    pub fn set_matches(&mut self, args: &'a [String]) -> Result<()> {
+        let trimmed_args: Vec<&str> = args
```

Comment:
I think we don't need trimmed_args. I think we can do all the required checks without it and removing it could simplify the validation logic.

Reply:
Removed it.